// @aws-cdk/aws-events-targets/test/codepipeline/pipeline.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = require("@aws-cdk/assert");
const codepipeline = require("@aws-cdk/aws-codepipeline");
const events = require("@aws-cdk/aws-events");
const core_1 = require("@aws-cdk/core");
const targets = require("../../lib");
test('use codebuild project as an eventrule target', () => {
    // GIVEN
    const stack = new core_1.Stack();
    const pipeline = new codepipeline.Pipeline(stack, 'Pipeline');
    const srcArtifact = new codepipeline.Artifact('Src');
    const buildArtifact = new codepipeline.Artifact('Bld');
    pipeline.addStage({
        stageName: 'Source',
        actions: [new TestAction({
                actionName: 'Hello',
                category: codepipeline.ActionCategory.SOURCE,
                provider: 'x',
                artifactBounds: { minInputs: 0, maxInputs: 0, minOutputs: 1, maxOutputs: 1, },
                outputs: [srcArtifact]
            })]
    });
    pipeline.addStage({
        stageName: 'Build',
        actions: [new TestAction({
                actionName: 'Hello',
                category: codepipeline.ActionCategory.BUILD,
                provider: 'y',
                inputs: [srcArtifact],
                outputs: [buildArtifact],
                artifactBounds: { minInputs: 1, maxInputs: 1, minOutputs: 1, maxOutputs: 1, }
            })]
    });
    const rule = new events.Rule(stack, 'rule', {
        schedule: events.Schedule.expression('rate(1 minute)'),
    });
    // WHEN
    rule.addTarget(new targets.CodePipeline(pipeline));
    const pipelineArn = {
        "Fn::Join": ["", [
                "arn:",
                { Ref: "AWS::Partition" },
                ":codepipeline:",
                { Ref: "AWS::Region" },
                ":",
                { Ref: "AWS::AccountId" },
                ":",
                { Ref: "PipelineC660917D" }
            ]
        ]
    };
    // THEN
    assert_1.expect(stack).to(assert_1.haveResource('AWS::Events::Rule', {
        Targets: [
            {
                Arn: pipelineArn,
                Id: "Target0",
                RoleArn: { "Fn::GetAtt": ["PipelineEventsRole46BEEA7C", "Arn"] }
            }
        ]
    }));
    assert_1.expect(stack).to(assert_1.haveResource('AWS::IAM::Policy', {
        PolicyDocument: {
            Statement: [
                {
                    Action: "codepipeline:StartPipelineExecution",
                    Effect: "Allow",
                    Resource: pipelineArn,
                }
            ],
            Version: "2012-10-17"
        }
    }));
});
class TestAction {
    constructor(actionProperties) {
        this.actionProperties = actionProperties;
        // nothing to do
    }
    bind(_scope, _stage, _options) {
        return {};
    }
    onStateChange(_name, _target, _options) {
        throw new Error('onStateChange() is not available on MockAction');
    }
}
