// @aws-cdk/aws-route53/test/integ.route53.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ec2 = require("@aws-cdk/aws-ec2");
const cdk = require("@aws-cdk/core");
const lib_1 = require("../lib");
const app = new cdk.App();
const stack = new cdk.Stack(app, 'aws-cdk-route53-integ');
const vpc = new ec2.Vpc(stack, 'VPC', { maxAzs: 1 });
const privateZone = new lib_1.PrivateHostedZone(stack, 'PrivateZone', {
    zoneName: 'cdk.local', vpc
});
const publicZone = new lib_1.PublicHostedZone(stack, 'PublicZone', {
    zoneName: 'cdk.test'
});
const publicSubZone = new lib_1.PublicHostedZone(stack, 'PublicSubZone', {
    zoneName: 'sub.cdk.test'
});
publicZone.addDelegation(publicSubZone);
new lib_1.TxtRecord(privateZone, 'TXT', {
    zone: privateZone,
    recordName: '_foo',
    values: [
        'Bar!',
        'Baz?'
    ],
    ttl: cdk.Duration.minutes(1)
});
new lib_1.CnameRecord(stack, 'CNAME', {
    zone: privateZone,
    recordName: 'www',
    domainName: 'server'
});
new lib_1.ARecord(stack, 'A', {
    zone: privateZone,
    recordName: 'test',
    target: lib_1.AddressRecordTarget.fromIpAddresses('1.2.3.4', '5.6.7.8')
});
new lib_1.CaaAmazonRecord(stack, 'CaaAmazon', {
    zone: publicZone
});
new cdk.CfnOutput(stack, 'PrivateZoneId', { value: privateZone.hostedZoneId });
new cdk.CfnOutput(stack, 'PublicZoneId', { value: publicZone.hostedZoneId });
app.synth();
