// @aws-cdk/aws-route53-targets/test/integ.api-gateway-domain-name.js
#!/usr/bin/env node
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const apig = require("@aws-cdk/aws-apigateway");
const acm = require("@aws-cdk/aws-certificatemanager");
const lambda = require("@aws-cdk/aws-lambda");
const route53 = require("@aws-cdk/aws-route53");
const core_1 = require("@aws-cdk/core");
const targets = require("../lib");
class TestStack extends core_1.Stack {
    constructor(scope, id) {
        super(scope, id);
        const domainName = 'example.com';
        const certArn = 'arn:aws:acm:us-east-1:111111111111:certificate';
        const hostedZoneId = 'AAAAAAAAAAAAA';
        const handler = new lambda.Function(this, 'Handler', {
            code: lambda.Code.fromInline(`exports.handler = async () => {
        return {
          statusCode: '200',
          body: 'hello, world!'
        };
      };`),
            runtime: lambda.Runtime.NODEJS_8_10,
            handler: 'index.handler'
        });
        const certificate = acm.Certificate.fromCertificateArn(this, 'cert', certArn);
        const api = new apig.LambdaRestApi(this, 'api', {
            handler,
            domainName: {
                certificate,
                domainName,
                endpointType: apig.EndpointType.REGIONAL,
            }
        });
        const zone = route53.HostedZone.fromHostedZoneAttributes(this, 'hosted-zone', {
            zoneName: domainName,
            hostedZoneId
        });
        new route53.ARecord(this, 'Alias', {
            zone,
            target: route53.AddressRecordTarget.fromAlias(new targets.ApiGateway(api))
        });
    }
}
const app = new core_1.App();
new TestStack(app, 'aws-cdk-apigw-alias-integ');
app.synth();
