// @aws-cdk/aws-codebuild/test/test.project.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const ec2 = require("@aws-cdk/aws-ec2");
const iam = require("@aws-cdk/aws-iam");
const aws_s3_1 = require("@aws-cdk/aws-s3");
const cdk = require("@aws-cdk/core");
const codebuild = require("../lib");
module.exports = {
    'can use filename as buildspec'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        new codebuild.Project(stack, 'Project', {
            source: codebuild.Source.s3({
                bucket: new aws_s3_1.Bucket(stack, 'Bucket'),
                path: 'path',
            }),
            buildSpec: codebuild.BuildSpec.fromSourceFilename('hello.yml'),
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::CodeBuild::Project', {
            Source: {
                BuildSpec: 'hello.yml'
            }
        }));
        test.done();
    },
    'can use buildspec literal'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        new codebuild.Project(stack, 'Project', {
            buildSpec: codebuild.BuildSpec.fromObject({ phases: ['say hi'] })
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::CodeBuild::Project', {
            Source: {
                BuildSpec: "{\n  \"phases\": [\n    \"say hi\"\n  ]\n}",
            }
        }));
        test.done();
    },
    'must supply buildspec when using nosource'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        test.throws(() => {
            new codebuild.Project(stack, 'Project', {});
        }, /you need to provide a concrete buildSpec/);
        test.done();
    },
    'must supply literal buildspec when using nosource'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        test.throws(() => {
            new codebuild.Project(stack, 'Project', {
                buildSpec: codebuild.BuildSpec.fromSourceFilename('bla.yml'),
            });
        }, /you need to provide a concrete buildSpec/);
        test.done();
    },
    'GitHub source': {
        'has reportBuildStatus on by default'(test) {
            // GIVEN
            const stack = new cdk.Stack();
            // WHEN
            new codebuild.Project(stack, 'Project', {
                source: codebuild.Source.gitHub({
                    owner: 'testowner',
                    repo: 'testrepo',
                    cloneDepth: 3,
                })
            });
            // THEN
            assert_1.expect(stack).to(assert_1.haveResource('AWS::CodeBuild::Project', {
                Source: {
                    Type: "GITHUB",
                    Location: 'https://github.com/testowner/testrepo.git',
                    ReportBuildStatus: true,
                    GitCloneDepth: 3,
                }
            }));
            test.done();
        },
        'can explicitly set reportBuildStatus to false'(test) {
            // GIVEN
            const stack = new cdk.Stack();
            // WHEN
            new codebuild.Project(stack, 'Project', {
                source: codebuild.Source.gitHub({
                    owner: 'testowner',
                    repo: 'testrepo',
                    reportBuildStatus: false,
                })
            });
            // THEN
            assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::CodeBuild::Project', {
                Source: {
                    ReportBuildStatus: false,
                },
            }));
            test.done();
        },
        'can explicitly set webhook to true'(test) {
            // GIVEN
            const stack = new cdk.Stack();
            // WHEN
            new codebuild.Project(stack, 'Project', {
                source: codebuild.Source.gitHub({
                    owner: 'testowner',
                    repo: 'testrepo',
                    webhook: true,
                })
            });
            // THEN
            assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::CodeBuild::Project', {
                Triggers: {
                    Webhook: true,
                },
            }));
            test.done();
        },
    },
    'project with s3 cache bucket'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        new codebuild.Project(stack, 'Project', {
            source: codebuild.Source.s3({
                bucket: new aws_s3_1.Bucket(stack, 'SourceBucket'),
                path: 'path',
            }),
            cache: codebuild.Cache.bucket(new aws_s3_1.Bucket(stack, 'Bucket'), {
                prefix: "cache-prefix"
            })
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::CodeBuild::Project', {
            Cache: {
                Type: "S3",
                Location: {
                    "Fn::Join": [
                        "/",
                        [
                            {
                                "Ref": "Bucket83908E77"
                            },
                            "cache-prefix"
                        ]
                    ]
                }
            },
        }));
        test.done();
    },
    'project with local cache modes'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        new codebuild.Project(stack, 'Project', {
            source: codebuild.Source.s3({
                bucket: new aws_s3_1.Bucket(stack, 'Bucket'),
                path: 'path',
            }),
            cache: codebuild.Cache.local(codebuild.LocalCacheMode.CUSTOM, codebuild.LocalCacheMode.DOCKER_LAYER, codebuild.LocalCacheMode.SOURCE)
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::CodeBuild::Project', {
            Cache: {
                Type: "LOCAL",
                Modes: [
                    "LOCAL_CUSTOM_CACHE",
                    "LOCAL_DOCKER_LAYER_CACHE",
                    "LOCAL_SOURCE_CACHE"
                ]
            },
        }));
        test.done();
    },
    'project by default has no cache modes'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        new codebuild.Project(stack, 'Project', {
            source: codebuild.Source.s3({
                bucket: new aws_s3_1.Bucket(stack, 'Bucket'),
                path: 'path',
            }),
        });
        // THEN
        assert_1.expect(stack).to(assert_1.not(assert_1.haveResourceLike('AWS::CodeBuild::Project', {
            Cache: {}
        })));
        test.done();
    },
    'can use an imported Role for a Project within a VPC'(test) {
        const stack = new cdk.Stack();
        const importedRole = iam.Role.fromRoleArn(stack, 'Role', 'arn:aws:iam::1234567890:role/service-role/codebuild-bruiser-service-role');
        const vpc = new ec2.Vpc(stack, 'Vpc');
        new codebuild.Project(stack, 'Project', {
            source: codebuild.Source.gitHubEnterprise({
                httpsCloneUrl: 'https://mygithub-enterprise.com/myuser/myrepo',
            }),
            role: importedRole,
            vpc,
        });
        assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::CodeBuild::Project', {
        // no need to do any assertions
        }));
        test.done();
    },
};
