// @aws-cdk/aws-events-targets/test/codebuild/codebuild.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = require("@aws-cdk/assert");
const codebuild = require("@aws-cdk/aws-codebuild");
const events = require("@aws-cdk/aws-events");
const core_1 = require("@aws-cdk/core");
const targets = require("../../lib");
test('use codebuild project as an eventrule target', () => {
    // GIVEN
    const stack = new core_1.Stack();
    const project = new codebuild.PipelineProject(stack, 'MyProject');
    const rule = new events.Rule(stack, 'Rule', {
        schedule: events.Schedule.expression('rate(1 min)')
    });
    // WHEN
    rule.addTarget(new targets.CodeBuildProject(project));
    // THEN
    assert_1.expect(stack).to(assert_1.haveResource('AWS::Events::Rule', {
        Targets: [
            {
                Arn: {
                    "Fn::GetAtt": [
                        "MyProject39F7B0AE",
                        "Arn"
                    ]
                },
                Id: "Target0",
                RoleArn: {
                    "Fn::GetAtt": [
                        "MyProjectEventsRole5B7D93F5",
                        "Arn"
                    ]
                }
            }
        ]
    }));
    assert_1.expect(stack).to(assert_1.haveResource('AWS::IAM::Role', {
        AssumeRolePolicyDocument: {
            Statement: [
                {
                    Action: "sts:AssumeRole",
                    Effect: "Allow",
                    Principal: { Service: "events.amazonaws.com" }
                }
            ],
            Version: "2012-10-17"
        }
    }));
    assert_1.expect(stack).to(assert_1.haveResource('AWS::IAM::Policy', {
        PolicyDocument: {
            Statement: [
                {
                    Action: "codebuild:StartBuild",
                    Effect: "Allow",
                    Resource: {
                        "Fn::GetAtt": [
                            "MyProject39F7B0AE",
                            "Arn"
                        ]
                    }
                }
            ],
            Version: "2012-10-17"
        }
    }));
});
test('specifying event for codebuild project target', () => {
    // GIVEN
    const stack = new core_1.Stack();
    const project = new codebuild.PipelineProject(stack, 'MyProject');
    const rule = new events.Rule(stack, 'Rule', {
        schedule: events.Schedule.expression('rate(1 hour)')
    });
    // WHEN
    const eventInput = {
        buildspecOverride: 'buildspecs/hourly.yml'
    };
    rule.addTarget(new targets.CodeBuildProject(project, {
        event: events.RuleTargetInput.fromObject(eventInput)
    }));
    // THEN
    assert_1.expect(stack).to(assert_1.haveResource('AWS::Events::Rule', {
        Targets: [
            {
                Arn: {
                    'Fn::GetAtt': ['MyProject39F7B0AE', 'Arn']
                },
                Id: 'Target0',
                Input: JSON.stringify(eventInput),
                RoleArn: {
                    'Fn::GetAtt': ['MyProjectEventsRole5B7D93F5', 'Arn']
                }
            }
        ]
    }));
});
