// @aws-cdk/aws-ecr-assets/lib/image-asset.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assets = require("@aws-cdk/assets");
const cdk = require("@aws-cdk/core");
const cxapi = require("@aws-cdk/cx-api");
const fs = require("fs");
const path = require("path");
const adopted_repository_1 = require("./adopted-repository");
/**
 * An asset that represents a Docker image.
 *
 * The image will be created in build time and uploaded to an ECR repository.
 */
class DockerImageAsset extends cdk.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        // resolve full path
        const dir = path.resolve(props.directory);
        if (!fs.existsSync(dir)) {
            throw new Error(`Cannot find image directory at ${dir}`);
        }
        if (!fs.existsSync(path.join(dir, 'Dockerfile'))) {
            throw new Error(`No 'Dockerfile' found in ${dir}`);
        }
        const staging = new assets.Staging(this, 'Staging', {
            ...props,
            sourcePath: dir
        });
        this.directory = staging.stagedPath;
        this.sourceHash = staging.sourceHash;
        const imageNameParameter = new cdk.CfnParameter(this, 'ImageName', {
            type: 'String',
            description: `ECR repository name and tag asset "${this.node.path}"`,
        });
        const asset = {
            id: this.node.uniqueId,
            packaging: 'container-image',
            path: this.directory,
            sourceHash: this.sourceHash,
            imageNameParameter: imageNameParameter.logicalId,
            repositoryName: props.repositoryName,
            buildArgs: props.buildArgs
        };
        this.node.addMetadata(cxapi.ASSET_METADATA, asset);
        // Parse repository name and tag from the parameter (<REPO_NAME>@sha256:<TAG>)
        // Example: cdk/cdkexampleimageb2d7f504@sha256:72c4f956379a43b5623d529ddd969f6826dde944d6221f445ff3e7add9875500
        const components = cdk.Fn.split('@sha256:', imageNameParameter.valueAsString);
        const repositoryName = cdk.Fn.select(0, components).toString();
        const imageSha = cdk.Fn.select(1, components).toString();
        // Require that repository adoption happens first, so we route the
        // input ARN into the Custom Resource and then get the URI which we use to
        // refer to the image FROM the Custom Resource.
        //
        // If adoption fails (because the repository might be twice-adopted), we
        // haven't already started using the image.
        this.repository = new adopted_repository_1.AdoptedRepository(this, 'AdoptRepository', { repositoryName });
        this.imageUri = `${this.repository.repositoryUri}@sha256:${imageSha}`;
        this.artifactHash = imageSha;
    }
}
exports.DockerImageAsset = DockerImageAsset;
