// @aws-cdk/aws-apigateway/test/test.http.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const cdk = require("@aws-cdk/core");
const apigateway = require("../lib");
module.exports = {
    'minimal setup'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        const api = new apigateway.RestApi(stack, 'my-api');
        // WHEN
        const integ = new apigateway.HttpIntegration('http://foo/bar');
        api.root.addMethod('GET', integ);
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('AWS::ApiGateway::Method', {
            Integration: {
                IntegrationHttpMethod: "GET",
                Type: "HTTP_PROXY",
                Uri: "http://foo/bar"
            }
        }));
        test.done();
    },
    'options can be passed via props'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        const api = new apigateway.RestApi(stack, 'my-api');
        // WHEN
        const integ = new apigateway.HttpIntegration('http://foo/bar', {
            httpMethod: 'POST',
            proxy: false,
            options: {
                cacheNamespace: 'hey'
            }
        });
        api.root.addMethod('GET', integ);
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('AWS::ApiGateway::Method', {
            Integration: {
                CacheNamespace: "hey",
                IntegrationHttpMethod: "POST",
                Type: "HTTP",
                Uri: "http://foo/bar"
            }
        }));
        test.done();
    }
};
