// @aws-cdk/aws-servicediscovery/test/test.namespace.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const ec2 = require("@aws-cdk/aws-ec2");
const cdk = require("@aws-cdk/core");
const servicediscovery = require("../lib");
module.exports = {
    'HTTP namespace'(test) {
        const stack = new cdk.Stack();
        new servicediscovery.HttpNamespace(stack, 'MyNamespace', {
            name: 'foobar.com',
        });
        assert_1.expect(stack).toMatch({
            Resources: {
                MyNamespaceD0BB8558: {
                    Type: "AWS::ServiceDiscovery::HttpNamespace",
                    Properties: {
                        Name: "foobar.com"
                    }
                }
            }
        });
        test.done();
    },
    'Public DNS namespace'(test) {
        const stack = new cdk.Stack();
        new servicediscovery.PublicDnsNamespace(stack, 'MyNamespace', {
            name: 'foobar.com',
        });
        assert_1.expect(stack).toMatch({
            Resources: {
                MyNamespaceD0BB8558: {
                    Type: "AWS::ServiceDiscovery::PublicDnsNamespace",
                    Properties: {
                        Name: "foobar.com"
                    }
                }
            }
        });
        test.done();
    },
    'Private DNS namespace'(test) {
        const stack = new cdk.Stack();
        const vpc = new ec2.Vpc(stack, 'MyVpc');
        new servicediscovery.PrivateDnsNamespace(stack, 'MyNamespace', {
            name: 'foobar.com',
            vpc
        });
        assert_1.expect(stack).to(assert_1.haveResource('AWS::ServiceDiscovery::PrivateDnsNamespace', {
            Name: "foobar.com",
            Vpc: {
                Ref: "MyVpcF9F0CA6F"
            }
        }));
        test.done();
    }
};
