// @aws-cdk/aws-cloudfront/test/integ.cloudfront-s3.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const iam = require("@aws-cdk/aws-iam");
const s3 = require("@aws-cdk/aws-s3");
const cdk = require("@aws-cdk/core");
const cloudfront = require("../lib");
const app = new cdk.App();
const stack = new cdk.Stack(app, 'integ-cloudfront-s3');
const bucket = new s3.Bucket(stack, 'Bucket', { removalPolicy: cdk.RemovalPolicy.DESTROY });
const oai = new cloudfront.CfnCloudFrontOriginAccessIdentity(stack, 'OAI', {
    cloudFrontOriginAccessIdentityConfig: {
        comment: 'Allows CloudFront to reach to the bucket!',
    }
});
const dist = new cloudfront.CloudFrontWebDistribution(stack, 'Distribution', {
    originConfigs: [{
            behaviors: [{ isDefaultBehavior: true }],
            s3OriginSource: {
                s3BucketSource: bucket,
                originAccessIdentityId: oai.ref,
            },
        }]
});
bucket.addToResourcePolicy(new iam.PolicyStatement({
    actions: ['s3:Get*', 's3:List*'],
    resources: [bucket.bucketArn, bucket.arnForObjects('*')],
    principals: [new iam.CanonicalUserPrincipal(oai.attrS3CanonicalUserId)],
}));
new cdk.CfnOutput(stack, 'DistributionDomainName', { value: dist.domainName });
