// @aws-cdk/aws-codecommit/test/test.codecommit.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const core_1 = require("@aws-cdk/core");
const lib_1 = require("../lib");
module.exports = {
    'CodeCommit Repositories': {
        'add an SNS trigger to repository'(test) {
            const stack = new core_1.Stack();
            const props = {
                repositoryName: 'MyRepository',
            };
            const snsArn = 'arn:aws:sns:*:123456789012:my_topic';
            new lib_1.Repository(stack, 'MyRepository', props).notify(snsArn);
            assert_1.expect(stack).toMatch({
                Resources: {
                    MyRepository4C4BD5FC: {
                        Type: "AWS::CodeCommit::Repository",
                        Properties: {
                            RepositoryName: "MyRepository",
                            Triggers: [
                                {
                                    Events: [
                                        "all"
                                    ],
                                    DestinationArn: "arn:aws:sns:*:123456789012:my_topic",
                                    Name: "MyRepository/arn:aws:sns:*:123456789012:my_topic"
                                }
                            ]
                        }
                    }
                }
            });
            test.done();
        },
        'fails when triggers have duplicate names'(test) {
            const stack = new core_1.Stack();
            const myRepository = new lib_1.Repository(stack, 'MyRepository', {
                repositoryName: 'MyRepository',
            }).notify('myTrigger');
            test.throws(() => myRepository.notify('myTrigger'));
            test.done();
        },
        'can be imported using a Repository ARN'(test) {
            // GIVEN
            const stack = new core_1.Stack();
            const repositoryArn = 'arn:aws:codecommit:us-east-1:585695036304:my-repo';
            // WHEN
            const repo = lib_1.Repository.fromRepositoryArn(stack, 'ImportedRepo', repositoryArn);
            // THEN
            test.deepEqual(stack.resolve(repo.repositoryArn), repositoryArn);
            test.deepEqual(stack.resolve(repo.repositoryName), 'my-repo');
            test.done();
        },
        'can be imported using just a Repository name (the ARN is deduced)'(test) {
            // GIVEN
            const stack = new core_1.Stack();
            // WHEN
            const repo = lib_1.Repository.fromRepositoryName(stack, 'ImportedRepo', 'my-repo');
            // THEN
            test.deepEqual(stack.resolve(repo.repositoryArn), {
                'Fn::Join': ['', [
                        'arn:',
                        { Ref: 'AWS::Partition' },
                        ':codecommit:',
                        { Ref: 'AWS::Region' },
                        ':',
                        { Ref: 'AWS::AccountId' },
                        ':my-repo'
                    ]],
            });
            test.deepEqual(stack.resolve(repo.repositoryName), 'my-repo');
            test.done();
        },
    },
};
