// @aws-cdk/aws-events-targets/test/aws-api/aws-api-handler.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const AWS = require("aws-sdk-mock");
const aws_api_handler_1 = require("../../lib/aws-api-handler");
console.log = jest.fn();
const event = {
    service: 'ECS',
    action: 'updateService',
    parameters: {
        service: 'cool-service',
        forceNewDeployment: true,
    },
};
beforeEach(() => {
    AWS.restore();
});
test('calls the SDK with the right parameters', async () => {
    const updateServiceMock = jest.fn().mockResolvedValue({ success: true });
    AWS.mock('ECS', 'updateService', updateServiceMock);
    await aws_api_handler_1.handler(event);
    expect(updateServiceMock).toHaveBeenCalledWith({
        service: 'cool-service',
        forceNewDeployment: true
    }, expect.any(Function));
    expect(console.log).toHaveBeenLastCalledWith('Response: %j', {
        success: true
    });
});
test('throws and logs in case of error', async () => {
    const updateServiceMock = jest.fn().mockRejectedValue({ code: 'Error' });
    AWS.mock('ECS', 'updateService', updateServiceMock);
    await expect(aws_api_handler_1.handler(event)).rejects.toEqual({ code: 'Error' });
    expect(console.log).toHaveBeenLastCalledWith({ code: 'Error' });
});
test('catches and logs error', async () => {
    const catchEvent = {
        ...event,
        catchErrorPattern: 'Invalid'
    };
    const updateServiceMock = jest.fn().mockRejectedValue({ code: 'Invalid' });
    AWS.mock('ECS', 'updateService', updateServiceMock);
    await aws_api_handler_1.handler(catchEvent);
    expect(console.log).toHaveBeenLastCalledWith({ code: 'Invalid' });
});
